include/alp/algorithms/householder_lu.hpp

```diff
@@ -0,0 +1,185 @@
+		 *
+		 * @tparam D        Data element type
+		 * @tparam Ring     Type of the semiring used in the computation
+		 * @tparam Minus    Type minus operator used in the computation
+		 * @tparam Divide   Type of divide operator used in the computation
+		 * @param[out] L    output lower trapezoidal matrix
+		 * @param[out] U    output upper trapezoidal matrix
+		 * @param[in]  H    input general matrix
+		 * @param[in]  ring A semiring for operations
+		 * @return RC       SUCCESS if the execution was correct
+		 *
+		 */
+		template<
+			typename MatH,
+			typename D = typename MatH::value_type,
+			typename MatL,
+			typename MatU,
+			class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			class Minus = operators::subtract< D >,
+			class Divide = operators::divide< D >,
+			std::enable_if_t<
+				is_matrix< MatH >::value &&
+				is_matrix< MatL >::value &&
+				is_matrix< MatU >::value &&
+				structures::is_a< typename MatH::structure, structures::General >::value &&
+				structures::is_a< typename MatL::structure, structures::LowerTrapezoidal >::value &&
+				structures::is_a< typename MatU::structure, structures::UpperTrapezoidal >::value &&
+				is_semiring< Ring >::value &&
+				is_operator< Minus >::value &&
+				is_operator< Divide >::value
+			> * = nullptr
+		>
+		RC householder_lu(
+			MatH &H,
+			MatL &L,
+			MatU &U,
+			Vector< size_t > &p,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+			RC rc = SUCCESS;
+
+			const Scalar< D > zero( ring.template getZero< D >() );
+			const Scalar< D > one( ring.template getOne< D >() );
+
+			const size_t m = nrows( H );
+			const size_t n = ncols( H );
+			const size_t k = std::min( n, m );
+
+			// check sizes
+			if(
+				( nrows( L ) != nrows( H ) ) ||
+				( ncols( U ) != ncols( H ) ) ||
+				( nrows( U ) != k ) ||
+				( ncols( L ) != k )
+			) {
+				std::cerr << " n, k, m = " << n << ", "  << k << ", " << m << "\n";
+				std::cerr << "Incompatible sizes in householder_lu.\n";
+				return FAILED;
+			}
+
+
+			// // L = identity( n )
+			auto Ldiag = alp::get_view< alp::view::diagonal >( L );
+			rc = rc ? rc : alp::set( Ldiag, one );
+			if( rc != SUCCESS ) {
+				std::cerr << " alp::set( L, I ) failed\n";
+				return rc;
+			}
+
+			// Out of place specification of the computation
+			MatH HWork( m, n );
+			rc = rc ? rc : alp::set( HWork, H );
+			if( rc != SUCCESS ) {
+				std::cerr << " alp::set( HWork, H ) failed\n";
+				return rc;
+			}
+
+			Vector< D > PivotVec( n );
+			rc = rc ? rc : alp::set( PivotVec, zero );
+
+			for( size_t k = 0; k < std::min( n, m ); ++k ) {
+				// =====   algorithm  =====
+				// a = H[ k, k ]
+				// v = H[ k + 1 : , k ]
+				// w = H[ k, k + 1 : ]
+				// Ak = H[ k + 1 :, k + 1 : ]
+				// v = v / a
+				// Ak = Ak - outer(v,w)
+				auto a_view = alp::get_view( HWork, utils::range( k, k + 1 ), k );
```

Comment:
This would be also resolved by foldl(Scalar, Matrix), which is an open issue.

Reply:
Just because of the time pressure, I would leave this as it is for now. Otherwise I agree, it should be resolved. @hidanielesgit  It this ok with you?